src/daft-core/src/array/ops/date.rs

```diff
@@ -108,6 +108,78 @@ impl TimestampArray {
         ))
     }
 
+    pub fn time(&self, timeunit_for_cast: &TimeUnit) -> DaftResult<TimeArray> {
+        let physical = self.physical.as_arrow();
+        let DataType::Timestamp(timeunit, tz) = self.data_type() else {
+            unreachable!("Timestamp array must have Timestamp datatype")
+        };
+        let tu = timeunit.to_arrow();
+        let timeunit_for_cast = if timeunit_for_cast == &TimeUnit::Nanoseconds {
+            TimeUnit::Nanoseconds
+        } else {
+            TimeUnit::Microseconds // default to microseconds
+        };
+        let time_arrow = match tz {
+            Some(tz) => match arrow2::temporal_conversions::parse_offset(tz) {
+                Ok(tz) => Ok(arrow2::array::PrimitiveArray::<i64>::from_iter(
+                    physical.iter().map(|ts| {
+                        ts.map(|ts| {
+                            let dt =
+                                arrow2::temporal_conversions::timestamp_to_datetime(*ts, tu, &tz);
+                            match timeunit_for_cast {
+                                TimeUnit::Nanoseconds => {
+                                    let hour = dt.hour() as i64 * 3_600_000_000_000;
+                                    let minute = dt.minute() as i64 * 60_000_000_000;
+                                    let second = dt.second() as i64 * 1_000_000_000;
+                                    let nanosecond = dt.nanosecond() as i64;
+                                    hour + minute + second + nanosecond
+                                }
+                                _ => {
+                                    let hour = dt.hour() as i64 * 3_600_000_000;
+                                    let minute = dt.minute() as i64 * 60_000_000;
+                                    let second = dt.second() as i64 * 1_000_000;
+                                    let microsecond = dt.nanosecond() as i64 / 1_000;
+                                    hour + minute + second + microsecond
+                                }
+                            }
```

Comment:
You might be able to accomplish this more simply by [converting](https://docs.rs/chrono/latest/chrono/struct.DateTime.html#method.time) the `DateTime` to the time (`NaiveTime`) component, [subtracting](https://docs.rs/chrono/latest/chrono/naive/struct.NaiveTime.html#method.signed_duration_since) `NaiveTime::from_hms(0, 0, 0)` to get a `TimeDelta` since midnight, and then using [`TimeDelta::num_microseconds()`](https://docs.rs/chrono/latest/chrono/struct.TimeDelta.html#method.num_microseconds) to get the number of microseconds since midnight (similar for nanoseconds).
```suggestion
                            let time_delta = dt.time() - NaiveTime::from_hms(0, 0, 0)
                            match timeunit_for_cast {
                                TimeUnit::Microseconds => time_delta.num_microseconds().unwrap(),
                                TimeUnit::Nanoseconds => time_delta.num_nanoseconds().unwrap(),
                                _ => unreachable!("Only microseconds and nanoseconds time units are supported for the Time dtype, but got {timeunit_for_cast}"),
                            }
```

Reply:
That's so much cleaner, thanks!